File: agenttel-python/src/agenttel/genai/anthropic.py
"""Anthropic SDK instrumentation wrapper."""

from __future__ import annotations

import functools
from typing import Any

from opentelemetry import trace

from agenttel.genai.span_builder import GenAiSpanBuilder


def instrument_anthropic(client: Any = None) -> Any:
    """Instrument an Anthropic client for automatic tracing.

    Wraps messages.create() to create gen_ai chat spans.

    Args:
        client: An anthropic.Anthropic client instance.

    Returns:
        The instrumented client.
    """
    if client is None:
        return client

    span_builder = GenAiSpanBuilder()
    _wrap_messages(client, span_builder)
    return client


def _wrap_messages(client: Any, span_builder: GenAiSpanBuilder) -> None:
    """Wrap the messages.create method."""
    if not hasattr(client, "messages"):
        return

    original_create = client.messages.create

    @functools.wraps(original_create)
    def traced_create(*args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        span = span_builder.start_chat_span(
            model=model,
            system="anthropic",
            temperature=kwargs.get("temperature"),
            max_tokens=kwargs.get("max_tokens"),
            top_p=kwargs.get("top_p"),
        )

        try:
            with trace.use_span(span, end_on_exit=False):
                response = original_create(*args, **kwargs)

                # Handle streaming
                if kwargs.get("stream"):
                    return _wrap_anthropic_stream(response, span, span_builder, model)

                _finalize_anthropic_span(span, span_builder, response, model)
                return response
        except Exception as exc:
            span.set_status(trace.Status(trace.StatusCode.ERROR, str(exc)))
            span.record_exception(exc)
            span.end()
            raise

    client.messages.create = traced_create


def _finalize_anthropic_span(
    span: trace.Span,
    span_builder: GenAiSpanBuilder,
    response: Any,
    model: str,
) -> None:
    """Extract Anthropic response data and end span."""
    input_tokens = 0
    output_tokens = 0
    cached_tokens = 0
    finish_reason = None
    response_id = None
    response_model = model

    if hasattr(response, "usage"):
        usage = response.usage
        input_tokens = getattr(usage, "input_tokens", 0) or 0
        output_tokens = getattr(usage, "output_tokens", 0) or 0
        cached_tokens = getattr(usage, "cache_read_input_tokens", 0) or 0

    if hasattr(response, "id"):
        response_id = response.id
    if hasattr(response, "model"):
        response_model = response.model
    if hasattr(response, "stop_reason"):
        finish_reason = response.stop_reason

    span_builder.end_span_with_response(
        span=span,
        model=response_model,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        cached_tokens=cached_tokens,
        finish_reason=finish_reason,
        response_id=response_id,
    )


def _wrap_anthropic_stream(
    stream: Any,
    span: trace.Span,
    span_builder: GenAiSpanBuilder,
    model: str,
) -> Any:
    """Wrap an Anthropic streaming response."""
    input_tokens = 0
    output_tokens = 0
    finish_reason = None
    response_model = model

    def _gen():
        nonlocal input_tokens, output_tokens, finish_reason, response_model
        try:
            for event in stream:
                event_type = getattr(event, "type", "")
                if event_type == "message_start" and hasattr(event, "message"):
                    msg = event.message
                    if hasattr(msg, "model"):
                        response_model = msg.model
                    if hasattr(msg, "usage"):
                        input_tokens = getattr(msg.usage, "input_tokens", 0) or 0
                elif event_type == "message_delta":
                    if hasattr(event, "usage"):
                        output_tokens = getattr(event.usage, "output_tokens", 0) or 0
                    if hasattr(event, "delta"):
                        finish_reason = getattr(event.delta, "stop_reason", None)
                yield event
        finally:
            span_builder.end_span_with_response(
                span=span,
                model=response_model,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                finish_reason=finish_reason,
            )

    return _gen()

File: agenttel-node/src/genai/anthropic.ts
/**
 * Anthropic SDK instrumentation wrapper.
 *
 * Monkey-patches `client.messages.create()` to automatically create
 * OpenTelemetry spans with GenAI semantic attributes for every message request.
 *
 * Usage:
 *   import Anthropic from '@anthropic-ai/sdk';
 *   import { instrumentAnthropic } from '@agenttel/node';
 *
 *   const client = instrumentAnthropic(new Anthropic());
 *   const msg = await client.messages.create({ model: 'claude-sonnet-4-20250514', ... });
 */

import { type Span } from '@opentelemetry/api';
import { GenAiSpanBuilder, type ChatResult } from './span-builder.js';

const SYSTEM = 'anthropic';

/**
 * Instrument an Anthropic client for automatic tracing.
 *
 * Wraps `messages.create()` to create gen_ai chat spans with full attribute
 * instrumentation including token usage, cost, cache metrics, and streaming.
 *
 * @param client - An Anthropic client instance.
 * @returns The instrumented client (same reference, mutated in place).
 */
export function instrumentAnthropic<T extends Record<string, any>>(client: T): T {
  if (!client) return client;

  const messages = client.messages;
  if (!messages?.create) return client;

  const spanBuilder = new GenAiSpanBuilder();
  const originalCreate = messages.create.bind(messages);

  messages.create = async function tracedCreate(...args: any[]) {
    const params = args[0] ?? {};
    const model: string = params.model ?? 'unknown';

    const span = spanBuilder.startChatSpan(model, SYSTEM);

    // Set optional request parameters.
    GenAiSpanBuilder.setRequestParams(
      span,
      params.temperature ?? undefined,
      params.max_tokens ?? undefined,
      params.top_p ?? undefined,
    );

    try {
      const response = await originalCreate(...args);

      // Handle streaming responses.
      if (params.stream) {
        return wrapStream(response, span, model);
      }

      // Non-streaming: extract response data and finalize span.
      finalizeSpan(span, response, model);
      return response;
    } catch (err) {
      GenAiSpanBuilder.endSpanError(span, err instanceof Error ? err : new Error(String(err)));
      throw err;
    }
  };

  return client;
}

/**
 * Extract Anthropic response data and end the span.
 */
function finalizeSpan(span: Span, response: any, requestModel: string): void {
  let inputTokens = 0;
  let outputTokens = 0;
  let cacheReadTokens = 0;
  let cacheCreateTokens = 0;
  let finishReason: string | undefined;
  let responseId: string | undefined;
  let responseModel = requestModel;

  if (response.usage) {
    inputTokens = response.usage.input_tokens ?? 0;
    outputTokens = response.usage.output_tokens ?? 0;
    cacheReadTokens = response.usage.cache_read_input_tokens ?? 0;
    cacheCreateTokens = response.usage.cache_creation_input_tokens ?? 0;
  }

  if (response.id) responseId = response.id;
  if (response.model) responseModel = response.model;
  if (response.stop_reason) finishReason = response.stop_reason;

  // Set cache token details.
  if (cacheReadTokens > 0 || cacheCreateTokens > 0) {
    GenAiSpanBuilder.setTokenDetails(span, {
      cachedReadTokens: cacheReadTokens || undefined,
      cachedWriteTokens: cacheCreateTokens || undefined,
    });
  }

  const result: ChatResult = {
    responseModel,
    responseId: responseId ?? '',
    inputTokens,
    outputTokens,
    finishReasons: finishReason ? [finishReason] : undefined,
  };

  GenAiSpanBuilder.endSpanSuccess(span, result);
}

/**
 * Wrap an Anthropic streaming response to accumulate token usage from
 * SSE events (message_start, content_block_delta, message_delta) and
 * end the span when the stream completes.
 */
function wrapStream(stream: any, span: Span, requestModel: string): AsyncIterable<any> {
  let inputTokens = 0;
  let outputTokens = 0;
  let cacheReadTokens = 0;
  let cacheCreateTokens = 0;
  let finishReason: string | undefined;
  let responseModel = requestModel;
  let responseId = '';

  const originalIterator = stream[Symbol.asyncIterator]
    ? stream[Symbol.asyncIterator].bind(stream)
    : null;

  if (!originalIterator) {
    GenAiSpanBuilder.endSpanSuccess(span, {
      responseModel,
      responseId: '',
      inputTokens: 0,
      outputTokens: 0,
    });
    return stream;
  }

  const wrappedIterable: AsyncIterable<any> = {
    [Symbol.asyncIterator]() {
      const iterator = originalIterator();
      return {
        async next() {
          try {
            const result = await iterator.next();
            if (result.done) {
              // Stream complete: finalize span with accumulated data.
              if (cacheReadTokens > 0 || cacheCreateTokens > 0) {
                GenAiSpanBuilder.setTokenDetails(span, {
                  cachedReadTokens: cacheReadTokens || undefined,
                  cachedWriteTokens: cacheCreateTokens || undefined,
                });
              }

              GenAiSpanBuilder.endSpanSuccess(span, {
                responseModel,
                responseId,
                inputTokens,
                outputTokens,
                finishReasons: finishReason ? [finishReason] : undefined,
              });
              return result;
            }

            const event = result.value;
            accumulateEvent(event);
            return result;
          } catch (err) {
            GenAiSpanBuilder.endSpanError(
              span,
              err instanceof Error ? err : new Error(String(err)),
            );
            throw err;
          }
        },
      };
    },
  };

  function accumulateEvent(event: any): void {
    const eventType = event.type ?? '';

    switch (eventType) {
      case 'message_start': {
        const msg = event.message;
        if (msg) {
          if (msg.model) responseModel = msg.model;
          if (msg.id) responseId = msg.id;
          if (msg.usage) {
            inputTokens = msg.usage.input_tokens ?? 0;
            cacheReadTokens = msg.usage.cache_read_input_tokens ?? 0;
            cacheCreateTokens = msg.usage.cache_creation_input_tokens ?? 0;
          }
        }
        break;
      }

      case 'message_delta': {
        if (event.usage) {
          outputTokens = event.usage.output_tokens ?? outputTokens;
        }
        if (event.delta?.stop_reason) {
          finishReason = event.delta.stop_reason;
        }
        break;
      }

      // content_block_start, content_block_delta, content_block_stop, message_stop
      // are passed through without accumulation.
    }
  }

  // Preserve any non-iterator properties from the original stream.
  return Object.assign(wrappedIterable, stream);
}
